rust/kernel/sync/arc.rs

```diff
@@ -49,8 +54,19 @@ unsafe impl<T: RefCounted + ?Sized + Sync + Send> Sync for Ref<T> {}
 impl<T: RefCounted> Ref<T> {
     /// Constructs a new reference counted instance of `T`.
     pub fn try_new(contents: T) -> Result<Self> {
-        let boxed = Box::try_new(contents)?;
-        boxed.get_count().count.store(1, Ordering::Relaxed);
+        Self::try_new_and_init(contents, |_| {})
+    }
+
+    /// Constructs a new reference counted instance of `T` and calls the initialisation function.
+    ///
+    /// This is useful because it provides a mutable reference to `T` at its final location.
+    pub fn try_new_and_init<U: FnOnce(&mut T)>(contents: T, init: U) -> Result<Self> {
+        let mut boxed = Box::try_new(contents)?;
+        // INVARIANT: The refcount is initialised to a non-zero value.
+        // SAFETY: We just allocated the object, no other references exist to it, so it is safe to
+        // initialise the refcount to 1.
+        unsafe { rust_helper_refcount_set(boxed.get_count().count.get(), 1) };
```

Comment:
I know that, in kernel C, lots of code is initializing `refcount_t` by simply zero-ing the memory, followed by a `refcount_set(n)`. Yes, it'll work.

But, is `REFCOUNT_INIT(n)` more correct? Initialization is not identical to setting an initial value, even if it happens to work for now?

Reply:
Yes, I wanted to use `REFCOUNT_INIT` but the fact that it's an rvalue turned me off.

But since we need a helper routine anyway for `refcount_set`, we may as well replace with a `refcount_init` that just returns a value initialised with `REFCOUNT_INIT`. Then we call it in the constructor of `RefCount`.

Do you have anything better in mind?

Reply:
Sounds good to me. In that case you won't even have to zero-init the structure using `default()`. And `#[derive(Default)]` on `RefCount` can go too.

Reply:
> Sounds good to me. In that case you won't even have to zero-init the structure using `default()`. And `#[derive(Default)]` on `RefCount` can go too.

That's what I was planning, but we can't really trust the value in `RefCount` that is passed into `try_new_and_init` -- for example, if the caller sets the value to 0, we would be breaking the invariant. We need to initialise it again in `try_new_and_init`.

Reply:
Maybe we cab add the zero-refcount requirement to the safety requirement of `RefCounted`?

Reply:
Yes, I see the problem. I'm not sure why `Ref<T>` requires its inner type to hold a `RefCount`, instead of working like `Arc<T>`? I'm sure there's a reason, I just don't know what it is.

Reply:
> Maybe we cab add the zero-refcount requirement to the safety requirement of `RefCounted`?

I suppose we could. But given the fact that we *can* enforce this, I think it's better to just reinit the refcount in `try_new_and_init` and contain the safety requirements.

Reply:
> Yes, I see the problem. I'm not sure why `Ref<T>` requires its inner type to hold a `RefCount`, instead of working like `Arc<T>`? I'm sure there's a reason, I just don't know what it is.

The reason for embedding the refcount is so that we can go from `&T` to `Ref<T>`, that is, we can increment the refcount on an object while having just a regular reference to it. We use it in Binder to increment the refcount on a `Process` object that implements `FileOperations`: in those functions we only get a `&Process`, using `Arc` we wouldn't have a safe path to go from `&Process` to `Arc<Process>`, which is need to do in some operations.

(We can actually build an `Arc<T>`-like abstractions from this. Perhaps we should use that instead to deprecate `Arc`.)

In my current design for red-black trees, we also only get `&T` when looking up elements stored as `Box<T>`, `Arc<T>` or `Ref<T>`, which is problematic for `Arc` when we need to take an extra reference.

Reply:
> I suppose we could. But given the fact that we _can_ enforce this, I think it's better to just reinit the refcount in `try_new_and_init` and contain the safety requirements.

Well, if we later change `RefCounted` to support C-manaaged pointers with `inc_ref/dec_ref` then this'll not be enforcable?